### PR TITLE
Auto tooltip position

### LINKED ISSUE #105: Some findings on tooltip
Hi just want to report some issues that I found when using o-tooltip

1. The tooltip positioning is not reacted to the edge

For example, the placement is top. Usually Bootstrap tooltip will auto adjust itself to Bottom to avoid the edge

image

------------

2. There is no variable to customize tooltip arrow color. For example the sceenshot show the tooltip arrow is using default border  color in my project

image

So if I customize the tooltip background color inside `app.scss`

```
$tooltip-background-color: rgba(0, 0, 0, 0.75);
@import "@oruga-ui/oruga/src/scss/oruga-full.scss";
```

It will looks like this. Notice that the arrow color doesnt change

image

Thanks for the hardwork!

### PR DESCRIPTION
**Pull request by blm768**
Fixes #105

Basic testbench is [here](https://github.com/blm768/oruga-tooltip-test).

## Proposed Changes

- Tooltips' `position` prop now accepts the value `"auto"`.
- Auto tooltip positioning picks the side that maximizes how much of the tooltip is visible onscreen.
- To break ties, auto positioning favors the default tooltip position.

## Open Questions

- [x] This solution assumes an arrow size of zero rather pulling the value from CSS. Is that good enough given that the arrow size is generally small?
- [ ] Should we add a small margin around the edge of the viewport so the tooltip doesn't run right up against it?
- [x] Should `auto` be the default value for the `position` prop?

## To Do

- [x] Test on mobile browsers; they handle viewports differently when zooming.
- [x] Backport to Vue 2

### CODE REVIEW

**Comment by jtommy** [comment]
I'll review soon in order to answer to all open questions.

**Comment by blm768** [comment]
Any thoughts on the open questions? I'm still interested in getting this merged if it'll add value.

**Comment by jtommy** [comment]
You are completely right, after the new release I'll give a try

**Comment by jtommy** [comment]
> This solution assumes an arrow size of zero rather pulling the value from CSS. Is that good enough given that the arrow size is generally small?

Yes, I think it's right... if anything we can add a prop to customize it (low priority)

>  Should we add a small margin around the edge of the viewport so the tooltip doesn't run right up against it?

Can you make an example?

> Should auto be the default value for the position prop?

Probably in the next breaking release

**Comment by blm768** [comment]
> > Should we add a small margin around the edge of the viewport so the tooltip doesn't run right up against it?
>
> Can you make an example?

Right now, we let the tooltip use its preferred position as long as it doesn't extend outside the viewport. In the first example in the image below, the tooltip is fully within the viewport, but it does run into the edge, so if it had a drop shadow, the shadow might be cut off by the viewport edge. If we consider that enough of a problem that we should do something about it, we could switch to an alternate position (shown here in gray) when the tooltip gets within, say, 5 pixels of the edge rather than at the point it actually crosses the edge.

example

> Are you able to support the Vue 2 version too?

Yep. I'm starting on that right now.

**Comment by blm768** [comment]
OK, I've put in the Vue 2 changes, and those seem to be working as expected. I also made a small change to how we compute the `isWebKit` flag so we can compensate for Edge's misbehavior.

**Comment by blm768** [comment]
I've also made the commit messages a little more descriptive.

**Review by jtommy — CHANGES_REQUESTED** [review]

**Review by jtommy — APPROVED** [review]

File: packages/oruga-next/src/components/tooltip/Tooltip.vue [hunk 1]
```diff
@@ -139,26 +186,87 @@ export default defineComponent({
...
+            if (this.position !== 'auto') {
+                return this.position
+            }
+            const defaultPosition = getValueByPath(getOptions(), 'tooltip.position', 'top')
+            let bestPosition = defaultPosition
+            if (this.metrics != null) {
+                let viewRect: DOMRect;
+                const viewport = (window as any).visualViewport; // Not available with our current types package
+                if (viewport != undefined) {
+                    console.debug("isWebKit: ", isWebKit);
```
[the remark this answers is not in the document]
Comment (blm768) [reply packages/oruga-next/src/components/tooltip/Tooltip.vue]: Oh, oops. Yeah, that was never supposed to actually get committed. I've pushed a separate commit that removes that, but I can squash that in with a quick rebase before this gets merged.

**Comment by blm768** [comment]
OK, fixups are squashed in.

**Comment by jtommy** [comment]
@blm768 thanks a lot! I'll add an example in the docs-next too

**Comment by jtommy** [comment]
@blm768 do you think we might use the sane approach on dropdown? It might be really helpful to add that feature to it and components that use it (datepicker, etc)

**Comment by blm768** [comment]
> Do you think we might use the same approach on dropdown?

I haven't looked too closely at the source yet, but I imagine it'd be possible to generalize this to work pretty much the same way with dropdowns. Depending on how we want them to behave, we might use different rules for repositioning them, and we might need to make more effort to handle long dropdowns (e.g. scroll to bring them onscreen, test cases where they're simply too long to fit onscreen) than we would for long tooltips just because dropdowns tend to run longer overall. At the very least, the check for whether the element is offscreen should be basically the same for both kinds of controls.

**Pull request merged**